File: ScreenCaptureNativeComponent/ImageHelper.cpp
#include "pch.h"
#include "ImageHelper.h"


IAsyncAction ImageHelper::SaveSoftwareBitmapToFileAsync(ABI::Windows::Graphics::Imaging::ISoftwareBitmap* pSoftwareBitmap, StorageFile const& outputFile)
{
	auto stream = co_await outputFile.OpenAsync(FileAccessMode::ReadWrite);
	auto encoder = co_await BitmapEncoder::CreateAsync(BitmapEncoder::JpegEncoderId(), stream);
	
	SoftwareBitmap bitmap{ nullptr };
	winrt::attach_abi(bitmap, pSoftwareBitmap);
	encoder.SetSoftwareBitmap(bitmap);
	encoder.IsThumbnailGenerated(true);

	try
	{
		co_await encoder.FlushAsync();
	}
	catch (...)
	{
		switch (winrt::to_hresult().value)
		{
		case WINCODEC_ERR_UNSUPPORTEDOPERATION:
			// If the encoder does not support writing a thumbnail, then try again
			// but disable thumbnail generation.
			encoder.IsThumbnailGenerated( false);
			break;
		default:
			throw;
		}
	}

	if (encoder.IsThumbnailGenerated() == false)
	{
		co_await encoder.FlushAsync();
	}
	pSoftwareBitmap->Release();
	stream.Close();
}

File: ScreenCapture/Helper/ImageHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.Graphics.Imaging;
using Windows.Media.Editing;
using Windows.Storage;
using Windows.Storage.Streams;

namespace ScreenCapture.Helper
{
    public class ImageHelper
    {
        public static async Task SaveSoftwareBitmapToFileAsync(SoftwareBitmap softwareBitmap, StorageFile outputFile)
        {
            using (IRandomAccessStream stream = await outputFile.OpenAsync(FileAccessMode.ReadWrite))
            {
                // Create an encoder with the desired format
                BitmapEncoder encoder = await BitmapEncoder.CreateAsync(BitmapEncoder.JpegEncoderId, stream);

                // Set the software bitmap
                encoder.SetSoftwareBitmap(softwareBitmap);
                encoder.IsThumbnailGenerated = false;
                await encoder.FlushAsync();
            }
        }
    }
}
